### DISCUSSION TITLE
Add support for storing large ipset's in RAM disks

### DISCUSSION THREAD

**Discussion by netravnen** [eworm-de/routeros-scripts#58] [Q&A] [answered]
Does it make sense to support storing large (>63000B) files in RAM disks before processing?

On Mikrotik CHR, CCR 2116, CCR 2216, RB 5009, RB 1100, CCR 2004, RB 3011 with >= 1G of RAM / Memory. It would make sense to allow sotring the fetched ipset files on a RAM disk before processing. This could alleviate the need to truncate the fetched ipset's.

```
warning: The list is huge and may be truncated: https://iplists.firehol.org/files/botscout_30d.ipset
warning: The list is huge and may be truncated: https://iplists.firehol.org/files/firehol_level2.netset
warning: The list is huge and may be truncated: https://iplists.firehol.org/files/firehol_level3.netset
warning: The list is huge and may be truncated: https://iplists.firehol.org/files/firehol_level4.netset
```
 
Currently, I have a ram-disk (1 GB) defined as follows,

```
/disk/add nfs-export=yes slot=tmp smb-export=yes smb-user=GUEST tmpfs-max-size=1073741824 type=tmpfs
```

on the following platform,

```
[admin@routeros] > /system/resource/print without-paging 
— cropped —
               board-name: CCR2116-12G-4S+
— cropped —
```

**Comment by eworm-de** [comment]
The limit is not defined by me, but RouterOS: The `fetch` command does not download more than ~ 64kB to a variable.

Actually it could download the complete list to a file, but that just moves the problem and makes it worse: Reading the content of a file is limited even more.

I have reported this to support (_SUP-132297_ "_fetch data truncated_"), and the answer was:

> It is an expected limit. We will note your request.

... and ...

> If there will be similar requests, we will consider increasing the output limit.

So feel free to send your own request...

**Reply by eworm-de** [reply]
Well, this is not what you expected or hoped for... Do you want to accept it as correct answer anyway?

**Comment by eworm-de** [comment]
Well, my comment made me think about it again, and I think I found a way to achieve support for "huge" downloads.

Want to give it a try? You need to [switch `global-functions` and `fw-addr-lists` to branch `staging`](https://git.eworm.de/cgit/routeros-scripts/about/BRANCHES.md).

This requires code from RouterOS 7.13, so it will possibly stay in `staging` for some time. Not yet sure when to bump...

**Reply by netravnen** [reply]
Hello @eworm-de, Sure! 👍 Will ~~set it up~~ update during this easter period and reply back here. 💯

**Reply by netravnen** [reply]
@eworm-de tried switching my [main] router at home to the staging branch for the two scripts you specified.

The temfs is created,

```
1 M   tmpfs  ram        5 693 112 320  5 693 046 784  tmpfs  none       
```

And some files are downloaded to the tmpfs partition

```
30  tmpfs/fw-addr-lists/$FetchHugeW7UKdHtyZfkbkozJuk4o           file              45        2024-04-02 21:22:03
31  tmpfs/fw-addr-lists/$FetchHugeiRarBPJBvtxvI5XkUpkc           file              45        2024-04-02 21:20:11
32  tmpfs                                                        disk                        2024-04-02 21:22:03
33  tmpfs/fw-addr-lists                                          directory                   2024-04-02 21:22:03
34  tmpfs/fw-addr-lists/$FetchHugeTABvvJtQa6cKMipazBYf           file              45        2024-04-02 21:13:47
35  tmpfs/fw-addr-lists/$FetchHugeydg9KxwrREvu6y4KFCPs           file              45        2024-04-02 21:18:53
```

After the files are downloaded, the script `fw-addr-lists` does not output any debug lines showing the firewall address lists are actually updated. `info: list: allow -- added: x - renewed: y - removed: z`

Current summary from my end is script `fw-addr-lists` **seems** to exit after having downloaded the lists.
- The lists are removed from /ip[v6,]/fireewall/address-list,
- no new entries are added. Essentially the script currently removes all configured entries for the address-lists
 
Will put the branch on a CHR during the week. To try debug a bit more what is happening. Using the list from https://iplists.firehol.org/?ipset=firehol_level4

**Reply by netravnen** [reply]
@eworm-de Would you be willing to leave the discussion open until a we confirm the suggested changes in branch staging is confirm working?

**Reply by eworm-de** [reply]
I did some small modifications... Can you update and try again, please?

What version of RouterOS are you running?

**Reply by eworm-de** [reply]
Oh, the "45" in your listing is the file size, no? 45 Bytes is not the huge, can you have a look at the content? Perhaps that gives a clue where problems may occur.

**Reply by netravnen** [reply]
> I did some small modifications... Can you update and try again, please?

Check. Updated the script version.

```
[admin@CCR] > system/script/print brief where comment="url-suffix=?h=staging"
Columns: NAME, OWNER, LAST-STARTED, RUN-COUNT
# NAME              OWNER             LAST-STARTED         RUN-COUNT
;;; url-suffix=?h=staging
2 global-functions  global-functions  2024-04-03 16:23:35         88
;;; url-suffix=?h=staging
7 fw-addr-lists     fw-addr-lists     2024-04-03 16:27:41        102
```

> What version of RouterOS are you running?

```
[admin@CCR] > /system/package/print proplist=name,version where name=routeros
Columns: NAME, VERSION
# NAME      VERSION
4 routeros  7.13  
```
> Oh, the "45" in your listing is the file size, no? 

```
[admin@CCR] > /file/print proplist=name,type,size,creation-time where name~"tmpfs"
Columns: NAME, TYPE, SIZE, CREATION-TIME
 #  NAME                                                 TYPE       SIZE  CREATION-TIME      
30  tmpfs/fw-addr-lists/-FetchHuge-zxMR00qH5rH0hShWNBqW  file        258  2024-04-03 16:27:42
31  tmpfs                                                disk             2024-04-03 16:26:23
32  tmpfs/fw-addr-lists                                  directory        2024-04-03 16:27:42
33  tmpfs/fw-addr-lists/-FetchHuge-ZK8rH96BeAa8UC15zmnp  file         45  2024-04-03 16:26:23
```

```
[admin@CCR] > file/read file=tmpfs/fw-addr-lists/-FetchHuge-ZK8rH96BeAa8UC15zmnp chunk-size=64k
value of chunk-size out of range (1..32768)

[admin@CCR] > file/read file=tmpfs/fw-addr-lists/-FetchHuge-ZK8rH96BeAa8UC15zmnp chunk-size=32768 offset=0
  data: 2405:9600::/32
                      2600:1400::/24
                                    2a02:26f0::/32

[admin@CCR] > file/read file=tmpfs/fw-addr-lists/-FetchHuge-zxMR00qH5rH0hShWNBqW chunk-size=32768 offset=0
  data: 2.16.0.0/13 23.0.0.0/12 23.32.0.0/11 23.64.0.0/14 23.72.0.0/13 23.192.0.0/11 69.192.0.0/16 72.246.0.0/15 88.221.0.0/16 92.122.0.0/15 95.100.0.0/15 
        96.6.0.0/15 96.16.0.0/15 104.64.0.0/10 118.214.0.0/16 173.222.0.0/15 184.24.0.0/13 184.50.0.0/15 184.84.0.0/14
```

**Reply by netravnen** [reply]
Looking over the fetch CMD

```
    /tool/fetch check-certificate=$CheckCert $Url dst=$FileName \
      http-header-field=({ [ $FetchUserAgentStr $ScriptName ] }) as-value;
```

When reading [the docs for fetch](https://help.mikrotik.com/docs/display/ROS/Fetch)... Does it make sense to change `dst` to `dst-path`?

**Reply by eworm-de** [reply]
That's crazy... Possibly a bug in `fetch` you see here? Any chance to update to version `7.13.1`? That comes with some fixes...

```
$DownloadPackage routeros 7.13.1;
/system/reboot;
```

(Oh, just scrolled up... You have other packages installed, no? You have to download all of them.)

**Reply by eworm-de** [reply]
> When reading [the docs for fetch](https://help.mikrotik.com/docs/display/ROS/Fetch)... Does it make sense to change `dst` to `dst-path`?

Yes, probably... Will change it later or tomorrow.

**Reply by netravnen** [reply]
I have all the required packages, afaik.
```
[admin@CCR] > /system/script/print brief proplist=name
Columns: NAME
 # NAME
 0 global-config
 1 global-config-overlay
;;; url-suffix=?h=staging
 2 global-functions
 3 check-certificates
 4 check-routeros-update
 5 check-wan-ip
 6 check-ggc-cluster
;;; url-suffix=?h=staging
 7 fw-addr-lists
 8 global-wait
 9 ipv6-update
10 mod/ssh-keys-import
11 mod/ipcalc
12 mod/inspectvar
13 packages-update
14 update-tunnelbroker
[admin@CCR] > /system/package/print proplist=name,version
Columns: NAME, VERSION
# NAME          VERSION
0 user-manager  7.13
1 iot           7.13
2 wireless      7.13
3 zerotier      7.13
4 routeros      7.13
5 rose-storage  7.13
6 container     7.13
[admin@CCR] > /file/print proplist=name,type,size,creation-time where name~"tmpfs"
Columns: NAME, TYPE, SIZE, CREATION-TIME
#  NAME                                                 TYPE       SIZE  CREATION-TIME
0  tmpfs/fw-addr-lists/-FetchHuge-zxMR00qH5rH0hShWNBqW  file        258  2024-04-03 16:27:42
1  tmpfs                                                disk             2024-04-03 16:26:23
2  tmpfs/fw-addr-lists                                  directory        2024-04-03 16:27:42
3  tmpfs/fw-addr-lists/-FetchHuge-ZK8rH96BeAa8UC15zmnp  file         45  2024-04-03 16:26:23
```

**Reply by eworm-de** [reply]
Oh, the scripts are fine. All required functions are provided in `global-functions`.

I was speaking about RouterOS and its optional packages. Any chance you update to version `7.13.1` and try then?

**Reply by netravnen** [reply]
Changing to 7.13.5 have so far not yielded much change...

```
[admin@CCR] > /system/script/print brief proplist=name
Flags: I - invalid
 #   NAME
 0   global-config
 1   global-config-overlay
 2   ;;; url-suffix=?h=staging
     global-functions
 3   check-certificates
 4   check-routeros-update
 5   check-wan-ip
 6   check-ggc-cluster
 7   ;;; url-suffix=?h=staging
     fw-addr-lists
 8   global-wait
 9   ipv6-update
10   mod/ssh-keys-import
11   mod/ipcalc
12   mod/inspectvar
13   packages-update
14   update-tunnelbroker
[admin@CCR] > /system/package/print proplist=name,version
Columns: NAME, VERSION
# NAME          VERSION
0 container     7.13.5
1 rose-storage  7.13.5
2 routeros      7.13.5
3 zerotier      7.13.5
4 wireless      7.13.5
5 iot           7.13.5
6 user-manager  7.13.5
[admin@CCR] > /file/print proplist=name,type,size,creation-time where name~"tmpfs"
Columns: NAME, TYPE, CREATION-TIME
#  NAME                 TYPE       CREATION-TIME
0  tmpfs                disk       2024-04-04 15:15:06
1  tmpfs/fw-addr-lists  directory  2024-04-04 15:21:12
```

**Reply by netravnen** [reply]
Afais, the lists are processed in alphabetical order.

`allow` and `akamai` are processed from my configuration file. `block` is never created. I.e. at takes _a very long times_ measured in bare minimum increments of ¼-hours. Currently the script is still holding a lock after running for 25 minutes.

```
[admin@CCR] > /system/script/print where name=global-config-overlay
Flags: I - invalid 
 1   name="global-config-overlay" owner="admin" policy=ftp,reboot,read,write,policy,test,password,sniff,sensitive,romon dont-require-permissions=no last-started=2024-04-04 15:07:35 run-count=67 
     source=
       # Overlay for global configuration by RouterOS Scripts
       # Copyright (c) 2013-2023 Christian Hesse <mail@eworm.de>
       # https://git.eworm.de/cgit/routeros-scripts/about/COPYING.md
       #
       # global configuration, custom overlay
       # https://git.eworm.de/cgit/routeros-scripts/about/#editing-configuration
       
       # Copy relevant configuration from global-config, paste and modify it here.
       # https://git.eworm.de/cgit/routeros-scripts/about/global-config.rsc
       
       # This defines the settings for firewall address-lists (fw-addr-lists).
       :global FwAddrLists {
         "allow"={
           { url="https://eworm.de/ros/fw-addr-lists/allow"; cert="E1"; timeout=1w };
         };
         "github-git"={
           { url="https://saasedl.paloaltonetworks.com/feeds/github/git/ipv6"; cert="GTS Root R1" };
           { url="https://saasedl.paloaltonetworks.com/feeds/github/git/ipv4"; cert="GTS Root R1" };
         };
         "block"={
           { url="https://eworm.de/ros/fw-addr-lists/block"; cert="E1" };
           { url="https://www.dshield.org/block.txt"; cidr="/24"; cert="R3" };
           { url="https://www.spamhaus.org/drop/drop.txt"; cert="Cloudflare Inc ECC CA-3" };
           { url="https://www.spamhaus.org/drop/edrop.txt"; cert="Cloudflare Inc ECC CA-3" };
           { url="https://iplists.firehol.org/files/firehol_level1.netset" };
           { url="https://iplists.firehol.org/files/firehol_level2.netset" };
           { url="https://iplists.firehol.org/files/firehol_level3.netset" };
           { url="https://iplists.firehol.org/files/firehol_level4.netset" };
           { url="https://iplists.firehol.org/files/botscout_7d.ipset" };
           { url="https://iplists.firehol.org/files/feodo.ipset" };
           { url="https://iplists.firehol.org/files/sslbl.ipset" };
           { url="https://iplists.firehol.org/files/zeus_badips.ipset" };
           { url="https://iplists.firehol.org/files/bambenek_c2.ipset" };
           { url="https://iplists.firehol.org/files/maxmind_proxy_fraud.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_apache.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_bots.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_bruteforce.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_ftp.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_imap.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_mail.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_sip.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_ssh.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_strongips.ipset" };
           { url="https://iplists.firehol.org/files/botvrij_dst.ipset" };
           { url="https://iplists.firehol.org/files/uscert_hidden_cobra.ipset" };
           { url="https://iplists.firehol.org/files/dyndns_ponmocup.ipset" };
           { url="https://iplists.firehol.org/files/cybercrime.ipset" };
           { url="https://iplists.firehol.org/files/iblocklist_malc0de.netset" };
           { url="https://iplists.firehol.org/files/turris_greylist.ipset" };
         };
         "microsoft-365"={
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/any/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/any/all/ipv4" };
         };
         "microsoft-teams"={
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/skype/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/skype/all/ipv4" };
         };
         "microsoft-sharepoint"={
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/sharepoint/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/sharepoint/all/ipv4" };
         };
         "microsoft-exchange"={
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/exchange/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/m365/worldwide/exchange/all/ipv4" };
         };
         "akamai"={
           { url="https://saasedl.paloaltonetworks.com/feeds/akamai/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/akamai/all/ipv4" };
         };
         "google-workspace"={
           { url="https://saasedl.paloaltonetworks.com/feeds/googleworkspace/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/googleworkspace/all/ipv4" };
         };
         "datadog"={
           { url="https://saasedl.paloaltonetworks.com/feeds/datadog/all/any/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/datadog/all/any/ipv4" };
         };
         "zoom"={
           { url="https://saasedl.paloaltonetworks.com/feeds/zoom/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/zoom/all/ipv4" };
         };
         "github-all"={
           { url="https://saasedl.paloaltonetworks.com/feeds/github/all/ipv6" };
           { url="https://saasedl.paloaltonetworks.com/feeds/github/all/ipv4" };
         };
       
       };
       :global FwAddrListTimeOut 3d;
       
       
       # End of global-config-overlay
```

**Reply by eworm-de** [reply]
Sure? The empty directory will stay, but there are no more download artifacts....

My guess is that address lists are populated. Did you check?

**Reply by netravnen** [reply]
Lost patience and upgraded 7.13.5 to 7.14.2

_Edit:_ from the 7.14 release notes

```
*) fetch - added "head" option for "http-method";
*) fetch - added "patch" option for "http-method";
*) fetch - allow specifying link-local address in FTP mode;
*) fetch - allow to use certificate and check-certificate parameters only in HTTPS mode;
*) fetch - do not require "content-length" for HTTP (introduced in v7.13);
*) fetch - fixed DNS resolving when domain has only AAAA entries (introduced in v7.13);
*) fetch - fixed fetch execution when unexpected data is received in HTTP payload;
*) fetch - fixed fetch when using "src-path" with HTTP/HTTPS modes (introduced in v7.13);
*) fetch - fixed fetch when using "src-path" with SFTP mode (introduced in v7.13);
*) fetch - fixed incorrect "src-path" error message when "upload=yes";
*) fetch - fixed IPv4 address logging (introduced in v7.13);
*) fetch - improved fetch stability in SFTP mode;
*) fetch - improved file download stability with HTTP/HTTPS modes;
*) fetch - less verbose logging;
*) fetch - print all "Set-Cookies" headers in response;
*) fetch - treat any 2xx HTTP return code as success (introduced in v7.13);
```

**Reply by netravnen** [reply]
Split the block list a bit, too. As I know the firehol_level1 (~2K), firehol_level2 (~19K), firehol_level3 (~18K) and firehol_level4 (~155K) lists are _very_ large.

```
         "firehol_level1"={
           { url="https://iplists.firehol.org/files/firehol_level1.netset" };
         };
         "firehol_level2"={
           { url="https://iplists.firehol.org/files/firehol_level2.netset" };
         };
         "firehol_level3"={
           { url="https://iplists.firehol.org/files/firehol_level3.netset" };
         };
         "firehol_level4"={
           { url="https://iplists.firehol.org/files/firehol_level4.netset" };
         };
         "block"={
           { url="https://eworm.de/ros/fw-addr-lists/block"; cert="E1" };
           { url="https://www.dshield.org/block.txt"; cidr="/24"; cert="R3" };
           { url="https://www.spamhaus.org/drop/drop.txt"; cert="Cloudflare Inc ECC CA-3" };
           { url="https://www.spamhaus.org/drop/edrop.txt"; cert="Cloudflare Inc ECC CA-3" };
           { url="https://iplists.firehol.org/files/botscout_7d.ipset" };
           { url="https://iplists.firehol.org/files/feodo.ipset" };
           { url="https://iplists.firehol.org/files/sslbl.ipset" };
           { url="https://iplists.firehol.org/files/zeus_badips.ipset" };
           { url="https://iplists.firehol.org/files/bambenek_c2.ipset" };
           { url="https://iplists.firehol.org/files/maxmind_proxy_fraud.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_apache.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_bots.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_bruteforce.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_ftp.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_imap.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_mail.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_sip.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_ssh.ipset" };
           { url="https://iplists.firehol.org/files/blocklist_de_strongips.ipset" };
           { url="https://iplists.firehol.org/files/botvrij_dst.ipset" };
           { url="https://iplists.firehol.org/files/uscert_hidden_cobra.ipset" };
           { url="https://iplists.firehol.org/files/dyndns_ponmocup.ipset" };
           { url="https://iplists.firehol.org/files/cybercrime.ipset" };
           { url="https://iplists.firehol.org/files/iblocklist_malc0de.netset" };
           { url="https://iplists.firehol.org/files/turris_greylist.ipset" };
         };
```

**Reply by netravnen** [reply]
Populating firehol_level2 (~19K) completed. firehol_level3 (~18K) completed. So far it has taken ~20 minutes to process the list.

I suspect firehol_level4 (~155K) will "kill" the script. In the sense the list is so large. And the script being limited to running on 1 of 16 cores (2 GHz clocked) is gonna take a while.

**Reply by netravnen** [reply]
Processing 18-19K entries takes around 3 minutes on a 2116 on a single 2GHz core. (firehol_level2, firehol_level3)

**Reply by netravnen** [reply]
**Sidenote:** It would be nice if the log message included the list name and the full url or a user configured name. With multiple lists from the same source domain. Only including the FQDN (the domain part) in the log message leads to needing educated guess-work for how far the script has reached.

```
"allow"={
  { name="eworm"; url="https://eworm.de/ros/fw-addr-lists/allow"; cert="E1"; timeout=1w };
};
```

E.g.
1. "Download in address-list allow list eworm FINISHED" (user configured name property)
2. "Download in address-list allow from https://eworm.de/ros/fw-addr-lists/allow FINISHED" (full url)

I prefer the example above that includes a user configured name property, with fall-back to using the full url if no name property has been configured.

<img width="448" alt="image" src="https://github.com/eworm-de/routeros-scripts/assets/1938389/d930c624-b810-4c7f-93b4-5d9b8f98c8fd">

**Reply by netravnen** [reply]
> That's crazy... Possibly a bug in fetch you see here? Any chance to update to version 7.13.1? That comes with some fixes...
> 
> ```
> $DownloadPackage routeros 7.13.1;
> /system/reboot;
> ```
> 
> (Oh, just scrolled up... You have other packages installed, no? You have to download all of them.)

Thank you for the tip. Forgot this command was included. :rocket: 🤘

```
:foreach pkgid in=[ /system/package/find ] do={
  :local pkgname [ /system/package/get $pkgid value-name=name ];
  $DownloadPackage $pkgname 7.14.2;
}
```

**Reply by eworm-de** [reply]
Note that the log messages you see are not from my script, but from `fetch` itself.

My script does log on download errors... I do not think "spamming" the log in normal operation makes sense. (However I do agree that it would help you to follow progress...)

**Reply by eworm-de** [reply]
Well, added a debug message... But for that to be shown or logged you need to take extra action.

Set this for debug messages in terminal:
```
:set PrintDebug true;
```
... or configure logging to include topic *debug* for messages from scripts:
```
/system/logging/add action=memory topics=debug,script;
```

**Reply by eworm-de** [reply]
For download of packages you can also use a script. 😜
https://git.eworm.de/cgit/routeros-scripts/about/doc/packages-update.md
... well, at least to go to latest version.

**Reply by eworm-de** [reply]
Regarding the changelog... Version `7.13.1` had a lot of `fetch` fixes as well, that's why I had wanted to test that first.

**Reply by eworm-de** [reply]
Just downgraded a CHR to 7.13, but could not reproduce any download issues. Possibly this always worked, but you did not give it the time to finish?

**Reply by eworm-de** [reply]
A final word on the overall situation: This is scripting, and limited to what RouterOS provides. Of course this is a lot slower than implementing in a "real" programming language - but this is not an option here. I've already optimized this as far as possible.

On the other side... Of course the initial import takes some time. But once populated you do not suffer any gaps, as lists are updated "in place". One cpu core is used for a long time on every invocation, though.

**Reply by netravnen** [reply]
In the end `firehol_level4.netset` finished downloading quite quickly. The proceesing of ~155K entries took forever with the processing being limited to 1 CPU core on 2116. Somewhere between 1 and 2 hours. Trying to extract logs from NMS system bogs it down. Will try to extract timestamps from the RAW MariaDB instance to get some timestamps.

**Reply by netravnen** [reply]
Printing the raw timestamps from the librenms syslog database. Processing ~155K entries of `firehol_level4.netset` took 2h40m.

```
MariaDB [librenms]> SELECT timestamp,msg
FROM syslog
WHERE device_id=5 AND program='DOWNLOAD' AND timestamp LIKE '2024-04-04%'
ORDER BY timestamp ASC
LIMIT 1000;
```

| timestamp (UTC) | msg
| --- | ---
| 2024-04-04 16:03:10 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 16:03:11 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 16:03:12 | from eworm.de FINISHED
| 2024-04-04 16:03:13 | from eworm.de FINISHED
| 2024-04-04 16:03:14 | from www.dshield.org FINISHED
| 2024-04-04 16:03:15 | from www.spamhaus.org FINISHED
| 2024-04-04 16:03:17 | from www.spamhaus.org FINISHED
| 2024-04-04 16:03:23 | from iplists.firehol.org FINISHED
| 2024-04-04 16:03:32 | from iplists.firehol.org FINISHED
| 2024-04-04 16:03:33 | from iplists.firehol.org FINISHED
| 2024-04-04 16:03:34 | from iplists.firehol.org FINISHED
| 2024-04-04 16:03:35 | from iplists.firehol.org FINISHED
| 2024-04-04 16:03:36 | from iplists.firehol.org FINISHED
| 2024-04-04 16:03:50 | from iplists.firehol.org FINISHED
| 2024-04-04 16:04:28 | from iplists.firehol.org FINISHED
| 2024-04-04 16:04:30 | from iplists.firehol.org FINISHED
| 2024-04-04 16:04:31 | from iplists.firehol.org FINISHED
| 2024-04-04 16:04:35 | from iplists.firehol.org FINISHED
| 2024-04-04 16:04:59 | from iplists.firehol.org FINISHED
| 2024-04-04 16:06:27 | from iplists.firehol.org FINISHED
| 2024-04-04 16:06:31 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:22 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:28 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:31 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:39 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:40 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:55 | from iplists.firehol.org FINISHED
| 2024-04-04 16:08:57 | from iplists.firehol.org FINISHED
| 2024-04-04 16:16:39 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 16:16:41 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 16:16:44 | from iplists.firehol.org FINISHED
| 2024-04-04 16:16:53 | from iplists.firehol.org FINISHED
| 2024-04-04 16:19:54 | from iplists.firehol.org FINISHED
| 2024-04-04 16:22:37 | from iplists.firehol.org FINISHED
| 2024-04-04 19:03:33 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:03:34 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:03:46 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:03:47 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:03:53 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:03:54 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:01 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:02 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:09 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:10 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:16 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:17 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:24 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:25 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:32 | from saasedl.paloaltonetworks.com FINISHED
| 2024-04-04 19:04:33 | from saasedl.paloaltonetworks.com FINISHED

**Reply by eworm-de** [reply]
Ok, so looks like works in general, but it is quite hard to import really huge lists.

Feel free to post more results and findings.

What still bothers me... Any idea if things worked with 7.13 already? (I think so...)

**Reply by netravnen** [reply]
> Just downgraded a CHR to 7.13, but could not reproduce any download issues. Possibly this always worked, but you did not give it the time to finish?

Correct.

**Reply by netravnen** [reply]
> What still bothers me... Any idea if things worked with 7.13 already? (I think so...)

Will retry 7.13.5 to se what times I get.

Isolating `firehol_level4.netset` to its own address-list assist in getting times for **huge** lists. Which I will use as a benchmark for the results going forwardens in this discussion.

**Reply by netravnen** [reply]
`:global PrintDebug true;` added this to `global-config-overlay` to see if I can get more accurate times in the log file.

**Reply by eworm-de** [reply]
Did another change to reduce the number of regular expression matches... Not sure if the change is noticeable.

Want to update and give it a try?

**Comment by eworm-de** [comment, marked as answer]
The new code made it through the `next` branch and into the `main` branch, and is available for anybody now. The size of a list is no longer limited by the script, but by devices' resources. (Also added a warning on this in documentation.)

**Reply by eworm-de** [reply]
(I think this is to be marked as correct answer...)

**Reply by netravnen** [reply]
Needed to change the category from Ideas to Q&A ([link](https://github.com/orgs/community/discussions/2903#discussioncomment-181405)) before being able to mark this as the answer.
